File: src/animallore/pace.ts
// The shard's own skill timer, learned rather than configured. Nothing in the API reports it, and a
// loop that guesses it low re-arms the very timer it is waiting out - which is what turns a fixed
// sleep into a run of nothing but 'you must wait'.
export const createPace = (options: {
  floor: number;
  step: number;
  max: number;
  easeAfter: number;
}) => {
  let delay = options.floor;
  let landed = 0;

  return {
    delay: (): number => delay,

    refused: (): number => {
      landed = 0;
      delay = Math.min(delay + options.step, options.max);

      return delay;
    },

    landed: (): number => {
      if (++landed < options.easeAfter) {
        return delay;
      }

      landed = 0;
      delay = Math.max(options.floor, delay - options.step);

      return delay;
    },
  };
};

File: legion/src/uo/pace.py
class Pace(object):
    """The shard's own skill timer, learned from its refusals rather than configured."""

    def __init__(self, floor, step, cap, ease_after):
        self._floor = floor
        self._step = step
        self._max = cap
        self._ease_after = ease_after
        self._delay = floor
        self._landed = 0

    def delay(self):
        return self._delay

    def refused(self):
        self._landed = 0
        self._delay = min(self._delay + self._step, self._max)

        return self._delay

    # Easing after a single success oscillates between an attempt and a refusal
    def landed(self):
        self._landed += 1

        if self._landed < self._ease_after:
            return self._delay

        self._landed = 0
        self._delay = max(self._floor, self._delay - self._step)

        return self._delay
